libraries/chain/host_context.cpp

```diff
@@ -39,54 +39,83 @@ host_context::~host_context() = default;
+         const auto max_depth = control.get_global_properties().configuration.max_sync_call_depth;
+         EOS_ASSERT(depth <= max_depth, sync_call_depth_exception,
+                    "reached sync call max call depth ${max_depth}", ("max_depth", max_depth));
+
+         const auto max_data_size = control.get_global_properties().configuration.max_sync_call_data_size;
+         EOS_ASSERT(data.size() <= max_data_size, sync_call_call_data_exception,
+                    "sync call call data size must be less or equal to ${max_data_size} bytes", ("max_data_size", max_data_size));
+
+         const auto* code = control.db().find<account_object, by_name>(call_receiver);
+         EOS_ASSERT(code != nullptr, sync_call_validate_exception,
+                    "sync call's receiver account ${r} does not exist", ("r", call_receiver));
+
+         EOS_ASSERT(flags <= static_cast<uint64_t>(sync_call_flags::all_allowed_bits), sync_call_validate_exception,  // all but `std::bit_width(all_allowed_bits)` LSBs must be 0s
+                    "only ${bits} least significant bits of sync call's flags (${flags}) can be set", ("bits", std::bit_width(static_cast<uint64_t>(sync_call_flags::all_allowed_bits)))("flags", flags));
+
          const account_metadata_object* receiver_account = &db.get<account_metadata_object, by_name>( call_receiver);
          if (receiver_account->code_hash.empty()) {
-            // TBD store the info in sync call trace
-            ilog("receiver_account->code_hash empty");
-            return return_value_size;
+            return handle_call_failure();
          }
 
          try {
             // use a new sync_call_context for next sync call
-            sync_call_context call_ctx(control, trx_context, get_sync_call_sender(), call_receiver, receiver_account->is_privileged(), depth, flags, data);
+            sync_call_context call_ctx(control, trx_context, ordinal, get_root_action_trace(), get_sync_call_sender(), call_receiver, receiver_account->is_privileged(), depth, flags, data);
 
+            if (control.contracts_console()) {
+               // only do this when console log is enabled; otherwise
+               // we will have a non-empty console markers vector with an empty console string
+               store_console_marker();
+            }
+
+            // execute the sync call
             auto rc = control.get_wasm_interface().do_sync_call(receiver_account->code_hash, receiver_account->vm_type, receiver_account->vm_version, call_ctx);
+
             if (rc == sync_call_return_code::receiver_not_support_sync_call) {  //  Currently -1 means there is no valid sync call entry point
-               return -1;
+               return handle_call_failure();
             }
 
             // store return value
             last_sync_call_return_value = std::move(call_ctx.return_value);
             return_value_size = last_sync_call_return_value.size();
          } catch( const wasm_exit&) {}
```

Comment:
Not new to this PR but I realize here that if a contract sets return value then `eosio_exit`s it looks like the return value is lost. Worse then that, highly likely there is a JIT vs OC difference here. Can make new issue for this but we need to track it some how if not fixed here.

Reply:
https://github.com/AntelopeIO/spring/issues/1372